### PR TITLE
fix: filterable multiselect readonly state implemented

### LINKED ISSUE #17587: [Bug]: Read Only state missing on `FilterableMultiselect` component
### Package

@carbon/react

### Browser

Chrome

### Package version

v11

### React version

v18

### Description

The implementation of read-only state for filterable multiselect is missing in the code. 

### Reproduction/example

https://react.carbondesignsystem.com/?path=/story/components-multiselect--filterable&args=readOnly:!false

### Steps to reproduce

1. Navigate to filterable story in the multiselect component.
2. Turn on the readOnly prop and check.

### Suggested Severity

Severity 2 = User cannot complete task, and/or no workaround within the user experience of a given component.

### Application/PAL

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by Gururajj77**
Closes #17587 

read only state for filterable multiselect is implemented

#### Changelog

**New**

- added new `readOnly` prop
- new `readOnlyEventHandlers` method to handle events during the `readOnly` state.

**Changed**

- changes related to readOnly state

#### Testing / Reviewing

- Go to the 'Filterable' story in the Multiselect component in the deploy preview
- Turn on the `readOnly` prop and see if it works as expected.

### CODE REVIEW

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx [hunk 1]
```diff
@@ -700,6 +709,14 @@ const FilterableMultiSelect = React.forwardRef(function FilterableMultiSelect<
...
       textInput.current?.focus();
     },
+    // onClick: (event) => {
+    //   if (!readOnly) {
+    //     handleMenuChange(!isOpen);
+    //     textInput.current?.focus();
+    //   } else {
+    //     event.preventDefault();
+    //   }
+    // },
```
Comment (preetibansalui) [inline packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: Please remove the commented code.

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx [hunk 2]
```diff
@@ -854,6 +898,8 @@ const FilterableMultiSelect = React.forwardRef(function FilterableMultiSelect<
             className={inputClasses}
             {...inputProps}
             ref={mergeRefs(textInput, inputProps.ref)}
```
Comment (preetibansalui) [inline packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: If its similar to const `mergedRef`, can we use that here?

**Review by preetibansalui — CHANGES_REQUESTED** [review]
Looks good, some minor changes only.

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx [hunk 3]
```diff
@@ -737,6 +754,11 @@ const FilterableMultiSelect = React.forwardRef(function FilterableMultiSelect<
           event.stopPropagation();
         }
 
+        if (readOnly) {
+          event.preventDefault();
+          return;
+        }
+
```
Comment (preetibansalui) [inline packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: I don't think code will reach here if we are using `readOnlyEventHandlers`.

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx (hunk 2 above)
Comment (Gururajj77) [reply packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: done

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx (hunk 3 above)
Comment (Gururajj77) [reply packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: removed

File: packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx [hunk 4]
```diff
@@ -728,7 +737,7 @@ const FilterableMultiSelect = React.forwardRef(function FilterableMultiSelect<
       placeholder,
       preventKeyAction: isOpen,
 
-      onClick: () => handleMenuChange(true),
+      onClick: () => !readOnly && handleMenuChange(true),
```
Comment (preetibansalui) [inline packages/react/src/components/MultiSelect/FilterableMultiSelect.tsx]: We are checking `readOnly` condition here and in `handleMenuChange` function both. I don't think it's necessary to have it in both places, as one check would also work same.

**Review by preetibansalui — CHANGES_REQUESTED** [review]
Thanks Guru for the changes, Sorry one minor change again. Otherwise it looks good to me.

**Review by preetibansalui — APPROVED** [review]
Thanks for the changes Guru, Looks Good now.

**Review by tay1orjones — COMMENTED** [review]
Looks good to me 👍 

@laurenmrice could you do a review as well? Just want to be sure we didn't miss anything design-wise

**Review by tay1orjones — APPROVED** [review]

**Review by laurenmrice — CHANGES_REQUESTED** [review]
- The tag border should be color token `$border-subtle`. I notice this is also wrong though in the regular Multiselect playground story.
Screenshot 2024-10-14 at 12 11 18 PM

---

- When you hover on the x icon in the tag, it turns black and lets you dismiss the tag. The x should not receive any color change or interaction in the read-only state.
Oct-14-2024 12-15-58

**Review by riddhybansal — APPROVED** [review]

**Comment by Gururajj77** [comment]
> * The tag border should be color token `$border-subtle`. I notice this is also wrong though in the regular Multiselect playground story.
> 
> Screenshot 2024-10-14 at 12 11 18 PM
> * When you hover on the x icon in the tag, it turns black and lets you dismiss the tag. The x should not receive any color change or interaction in the read-only state.
>   Oct-14-2024 12-15-58
>         
>           [
>             
>           
>               Oct-14-2024 12-15-58
>             ](https://private-user-images.githubusercontent.com/43969356/376318651-19bd49c1-165c-403f-9592-82bd2bae5d45.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5ODEyMjMsIm5iZiI6MTcyODk4MDkyMywicGF0aCI6Ii80Mzk2OTM1Ni8zNzYzMTg2NTEtMTliZDQ5YzEtMTY1Yy00MDNmLTk1OTItODJiZDJiYWU1ZDQ1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDA4Mjg0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlYmViOTc0MTNjNzhjYmQ0ODlmMWYzMjQ3MTZkM2Q4ZTVjNjQyOTFlOTBkMGUyOGEzZGY5YThiMGNjNGVmZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1_3_WcAYqMavjf-SQ7ktlncwXeJPTq6viUD3CN0hajg)
>           
>           
>             
>               
>                 
>               
>               
>                 
>                 
>               
>             
>             [
>               
>                 
>               
>             ](https://private-user-images.githubusercontent.com/43969356/376318651-19bd49c1-165c-403f-9592-82bd2bae5d45.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5ODEyMjMsIm5iZiI6MTcyODk4MDkyMywicGF0aCI6Ii80Mzk2OTM1Ni8zNzYzMTg2NTEtMTliZDQ5YzEtMTY1Yy00MDNmLTk1OTItODJiZDJiYWU1ZDQ1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDA4Mjg0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlYmViOTc0MTNjNzhjYmQ0ODlmMWYzMjQ3MTZkM2Q4ZTVjNjQyOTFlOTBkMGUyOGEzZGY5YThiMGNjNGVmZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1_3_WcAYqMavjf-SQ7ktlncwXeJPTq6viUD3CN0hajg)

Hey Lauren, I have added the functionality to not have any interaction during the `readOnly` state in FilterableMultiselect, I'll create new issues for the Multiselect interaction and the `$border-subtle` mismatch for both components.

**Review by tay1orjones — APPROVED** [review]

**Comment by laurenmrice** [comment]
Okay thanks, @Gururajj77 ! One last thing I noticed is if you hover on the number in the tag in the read-only state, a disabled mouse cursor appears. The mouse cursor should just be the arrow, like how it appears when you hover on the x icon in the tag. I also see the number 2 in the tag looks like it has a highlighted background color behind it which it should not have.

Oct-15-2024 15-02-29

**Review by laurenmrice — CHANGES_REQUESTED** [review]
Okay thanks, @Gururajj77 ! One last thing I noticed for the Filterable Multiselect is if you hover on the number in the tag in the read-only state, a disabled mouse cursor appears. The mouse cursor should just be the arrow, like how it appears when you hover on the x icon in the tag. I also see the number 2 in the tag looks like it has a highlighted background color behind it which it should not have.

Oct-15-2024 15-02-29

**Comment by Gururajj77** [comment]
Have created an issue, #17820 to keep track for the hover issue separately

**Review by laurenmrice — APPROVED** [review]

**Pull request merged**